File: assets/scripts/gen_galvanometer.py
"""
Generate a detailed 384x384 tangent galvanometer sprite.
PICO-8 palette, steampunk scientific instrument aesthetic.
High-res for clarity when scaled in-game.
"""
import math
from PIL import Image, ImageDraw, ImageFont

# PICO-8 palette
P = {
    'BLACK':       (0x00, 0x00, 0x00),
    'DEEP_NAVY':   (0x0d, 0x1b, 0x2a),
    'DARK_BLUE':   (0x1b, 0x28, 0x38),
    'MED_BLUE':    (0x2d, 0x4a, 0x6a),
    'BRIGHT_BLUE': (0x29, 0xad, 0xff),
    'CYAN':        (0x53, 0xd8, 0xfb),
    'WHITE':       (0xff, 0xf1, 0xe8),
    'LIGHT_GREY':  (0xc2, 0xc3, 0xc7),
    'DARK_GREY':   (0x5f, 0x57, 0x4f),
    'RED':         (0xff, 0x00, 0x4d),
    'ORANGE':      (0xff, 0xa3, 0x00),
    'YELLOW':      (0xff, 0xec, 0x27),
    'GREEN':       (0x00, 0xe4, 0x36),
    'PINK':        (0xff, 0x77, 0xa8),
    'INDIGO':      (0x83, 0x76, 0x9c),
    'BROWN':       (0xab, 0x52, 0x36),
    'SKIN':        (0xff, 0xcc, 0xaa),
}

SIZE = 384
CX, CY = SIZE // 2, SIZE // 2

img = Image.new('RGBA', (SIZE, SIZE), (0, 0, 0, 0))
draw = ImageDraw.Draw(img)

def circle(cx, cy, r, fill=None, outline=None, width=1):
    box = [cx - r, cy - r, cx + r, cy + r]
    if fill:
        draw.ellipse(box, fill=fill)
    if outline:
        draw.ellipse(box, outline=outline, width=width)

def polar(cx, cy, r, angle_deg):
    rad = math.radians(angle_deg)
    return (cx + r * math.cos(rad), cy - r * math.sin(rad))

def line_polar(cx, cy, r1, r2, angle_deg, color, width=1):
    x1, y1 = polar(cx, cy, r1, angle_deg)
    x2, y2 = polar(cx, cy, r2, angle_deg)
    draw.line([(x1, y1), (x2, y2)], fill=color, width=width)

# === MOUNTING PLATE CORNERS (behind everything) ===
plate_color = (*P['DARK_GREY'], 180)
draw.polygon([(CX-180, CY-180), (CX-140, CY-180), (CX-180, CY-140)], fill=plate_color)
draw.polygon([(CX+180, CY-180), (CX+140, CY-180), (CX+180, CY-140)], fill=plate_color)
draw.polygon([(CX-180, CY+180), (CX-140, CY+180), (CX-180, CY+140)], fill=plate_color)
draw.polygon([(CX+180, CY+180), (CX+140, CY+180), (CX+180, CY+140)], fill=plate_color)

# === COPPER COIL RING (outermost) ===
# Thick copper ring
circle(CX, CY, 186, fill=P['ORANGE'])
circle(CX, CY, 186, outline=P['BROWN'], width=3)
# Coil wire wrapping texture
for a in range(0, 360, 3):
    x1, y1 = polar(CX, CY, 176, a)
    x2, y2 = polar(CX, CY, 186, a)
    draw.line([(x1, y1), (x2, y2)], fill=P['BROWN'], width=2)
# Copper highlight (upper-left area for 3D effect)
for a in range(200, 320, 4):
    x, y = polar(CX, CY, 181, a)
    draw.rectangle([x-2, y-2, x+2, y+2], fill=(*P['YELLOW'], 90))
# Inner copper ring edge
circle(CX, CY, 172, outline=P['BROWN'], width=3)

# === BRASS BEZEL (outer metallic ring) ===
circle(CX, CY, 168, fill=P['DARK_GREY'])
circle(CX, CY, 168, outline=P['LIGHT_GREY'], width=3)
# Bezel highlights (top-left arc for 3D)
for a in range(120, 240, 2):
    x, y = polar(CX, CY, 165, a)
    draw.rectangle([x, y, x+2, y+2], fill=(*P['LIGHT_GREY'], 100))
# Bezel shadow (bottom-right arc)
for a in range(300, 420, 2):
    aa = a % 360
    x, y = polar(CX, CY, 165, aa)
    draw.rectangle([x, y, x+2, y+2], fill=(*P['BLACK'], 80))

# === INNER FACE (dark blue) ===
circle(CX, CY, 155, fill=P['DEEP_NAVY'])
circle(CX, CY, 155, outline=P['MED_BLUE'], width=2)

# === DEGREE MARKINGS ===
for deg in range(0, 360, 5):
    angle = 90 - deg  # 0 at top (north), clockwise
    if deg % 90 == 0:
        line_polar(CX, CY, 130, 155, angle, P['WHITE'], width=4)
    elif deg % 45 == 0:
        line_polar(CX, CY, 134, 155, angle, P['WHITE'], width=3)
    elif deg % 15 == 0:
        line_polar(CX, CY, 138, 155, angle, P['LIGHT_GREY'], width=3)
    elif deg % 10 == 0:
        line_polar(CX, CY, 141, 155, angle, P['LIGHT_GREY'], width=2)
    else:
        line_polar(CX, CY, 145, 155, angle, P['LIGHT_GREY'], width=1)

# === DEGREE NUMBERS (big, bold) ===
# Try fonts at different sizes
def get_font(size):
    for path in [
        "arial.ttf",
        "arialbd.ttf",
        "C:/Windows/Fonts/arialbd.ttf",
        "C:/Windows/Fonts/arial.ttf",
        "/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf",
        "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf",
    ]:
        try:
            return ImageFont.truetype(path, size)
        except:
            continue
    return ImageFont.load_default()

num_font = get_font(22)
big_font = get_font(34)

# For tangent galvanometer: scale -90 to +90
# 0 at top, 90 at east (right), -90 at west (left)
num_positions = [
    (0,   '0'),
    (30,  '30'),
    (45,  '45'),
    (60,  '60'),
    (90,  '90'),
    (330, '30'),     # -30 displayed as 30 (symmetric)
    (315, '45'),     # -45
    (300, '60'),     # -60
    (270, '90'),     # -90
]

for (deg, label) in num_positions:
    angle = 90 - deg
    tx, ty = polar(CX, CY, 118, angle)
    bbox = draw.textbbox((0, 0), label, font=num_font)
    tw, th = bbox[2] - bbox[0], bbox[3] - bbox[1]
    draw.text((tx - tw/2, ty - th/2), label, fill=P['WHITE'], font=num_font)

# === N and S LABELS (large, prominent) ===
# N at top (red) - inside the markings
bbox = draw.textbbox((0, 0), 'N', font=big_font)
nw, nh = bbox[2] - bbox[0], bbox[3] - bbox[1]
draw.text((CX - nw/2, CY - 142 - nh/2), 'N', fill=P['RED'], font=big_font)

# S at bottom (blue)
bbox = draw.textbbox((0, 0), 'S', font=big_font)
sw, sh = bbox[2] - bbox[0], bbox[3] - bbox[1]
draw.text((CX - sw/2, CY + 142 - sh/2), 'S', fill=P['BRIGHT_BLUE'], font=big_font)

# + and - signs on the sides
bbox = draw.textbbox((0, 0), '+', font=big_font)
pw, ph = bbox[2] - bbox[0], bbox[3] - bbox[1]
draw.text((CX + 92 - pw/2, CY - 68), '+', fill=P['YELLOW'], font=big_font)

bbox = draw.textbbox((0, 0), '-', font=big_font)
mw, mh = bbox[2] - bbox[0], bbox[3] - bbox[1]
draw.text((CX - 92 - mw/2, CY - 68), '-', fill=P['YELLOW'], font=big_font)

# === CONCENTRIC GUIDE RINGS ===
circle(CX, CY, 100, outline=(*P['MED_BLUE'], 50), width=1)
circle(CX, CY, 70, outline=(*P['MED_BLUE'], 35), width=1)

# === CROSS HAIRS (thin reference lines) ===
# Vertical center line
draw.line([(CX, CY - 95), (CX, CY - 10)], fill=(*P['MED_BLUE'], 40), width=1)
draw.line([(CX, CY + 10), (CX, CY + 95)], fill=(*P['MED_BLUE'], 40), width=1)
# Horizontal center line
draw.line([(CX - 95, CY), (CX - 10, CY)], fill=(*P['MED_BLUE'], 40), width=1)
draw.line([(CX + 10, CY), (CX + 95, CY)], fill=(*P['MED_BLUE'], 40), width=1)

# === CENTER PIVOT ===
circle(CX, CY, 10, fill=P['LIGHT_GREY'])
circle(CX, CY, 7, fill=P['WHITE'])
circle(CX, CY, 4, fill=(*P['YELLOW'], 220))
circle(CX, CY, 2, fill=P['WHITE'])

# === BRASS SCREWS at 4 mounting positions ===
screw_positions = [(CX-160, CY-160), (CX+160, CY-160), (CX-160, CY+160), (CX+160, CY+160)]
for sx, sy in screw_positions:
    if 8 < sx < SIZE-8 and 8 < sy < SIZE-8:
        circle(sx, sy, 8, fill=P['DARK_GREY'])
        circle(sx, sy, 8, outline=P['LIGHT_GREY'], width=2)
        draw.line([(sx-5, sy), (sx+5, sy)], fill=P['BLACK'], width=2)
        draw.rectangle([sx-2, sy-5, sx+1, sy-3], fill=(*P['WHITE'], 100))

# === GLASS REFLECTION HIGHLIGHTS ===
for i in range(25):
    x = CX - 60 + i * 3
    y = CY - 90 + int(math.sin(i * 0.25) * 12)
    w = max(1, 4 - i // 7)
    draw.rectangle([x, y, x+w, y+2], fill=(*P['WHITE'], 35 + i * 2))

circle(CX - 45, CY - 75, 5, fill=(*P['WHITE'], 50))

for i in range(12):
    x = CX - 35 + i * 3
    y = CY - 70 + i * 2
    draw.rectangle([x, y, x+2, y+1], fill=(*P['WHITE'], 25))

# Save
output_path = r'C:\Users\shlom\projects\DNA Game\assets\sprites\galvanometer.png'
img.save(output_path)
print(f'Saved galvanometer sprite to {output_path} ({img.size[0]}x{img.size[1]})')

File: assets/scripts/gen_galvanometer.lua
-- Tangent Galvanometer (192x192) - Detailed scientific instrument
local spr = Sprite(192, 192, ColorMode.RGB)
spr.filename = "galvanometer.png"
local cel = spr.cels[1]
local img = cel.image

-- PICO-8 Palette
local DEEP_NAVY = Color{r=13, g=27, b=42, a=255}
local DARK_BLUE = Color{r=27, g=40, b=56, a=255}
local MED_BLUE = Color{r=45, g=74, b=106, a=255}
local BRIGHT_BLUE = Color{r=41, g=173, b=255, a=255}
local CYAN = Color{r=83, g=216, b=251, a=255}
local WHITE = Color{r=255, g=241, b=232, a=255}
local LIGHT_GREY = Color{r=194, g=195, b=199, a=255}
local DARK_GREY = Color{r=95, g=87, b=79, a=255}
local RED = Color{r=255, g=0, b=77, a=255}
local ORANGE = Color{r=255, g=163, b=0, a=255}
local YELLOW = Color{r=255, g=236, b=39, a=255}
local BROWN = Color{r=171, g=82, b=54, a=255}
local TRANS = Color{r=0, g=0, b=0, a=0}

-- Brass tones
local BRASS = Color{r=200, g=170, b=80, a=255}
local BRASS_DARK = Color{r=140, g=110, b=40, a=255}
local BRASS_LIGHT = Color{r=230, g=205, b=120, a=255}
local BRASS_BRIGHT = Color{r=245, g=225, b=160, a=255}

local cx = 96
local cy = 96
local PI = 3.14159265

-- Clear to transparent
for y = 0, 191 do
  for x = 0, 191 do
    img:drawPixel(x, y, TRANS)
  end
end

-- Helper: distance
local function dist(x, y)
  return math.sqrt((x - cx)*(x - cx) + (y - cy)*(y - cy))
end

-- Helper: simple atan2 replacement
local function myatan2(dy, dx)
  if dx > 0 then return math.atan(dy / dx)
  elseif dx < 0 and dy >= 0 then return math.atan(dy / dx) + PI
  elseif dx < 0 and dy < 0 then return math.atan(dy / dx) - PI
  elseif dx == 0 and dy > 0 then return PI / 2
  elseif dx == 0 and dy < 0 then return -PI / 2
  else return 0 end
end

-- Helper: blend colors
local function blendC(c1, c2, t)
  if t < 0 then t = 0 end
  if t > 1 then t = 1 end
  return Color{
    r = math.floor(c1.red * (1-t) + c2.red * t),
    g = math.floor(c1.green * (1-t) + c2.green * t),
    b = math.floor(c1.blue * (1-t) + c2.blue * t),
    a = 255
  }
end

-- ── PASS 1: Draw all rings ──
for y = 0, 191 do
  for x = 0, 191 do
    local d = dist(x, y)
    local dx = x - cx
    local dy = y - cy
    local angle = myatan2(dy, dx)
    local lightFactor = math.cos(angle + PI * 0.75) * 0.5 + 0.5

    if d <= 94 and d > 88 then
      -- COPPER COIL RING with winding texture
      local windingPhase = math.sin(angle * 30)
      if windingPhase > 0.3 then
        img:drawPixel(x, y, blendC(ORANGE, YELLOW, 0.3))
      elseif windingPhase > -0.3 then
        img:drawPixel(x, y, ORANGE)
      else
        img:drawPixel(x, y, blendC(ORANGE, BROWN, 0.5))
      end

    elseif d <= 88 and d > 82 then
      -- BRASS BEZEL with metallic shading
      if d > 86 then
        img:drawPixel(x, y, blendC(BRASS_DARK, BRASS, lightFactor * 0.5))
      elseif d < 83 then
        img:drawPixel(x, y, blendC(BRASS_DARK, BRASS, lightFactor * 0.4))
      else
        img:drawPixel(x, y, blendC(BRASS_DARK, BRASS_BRIGHT, lightFactor))
      end

    elseif d <= 82 and d > 60 then
      -- SCALE RING area (dark)
      local normD = (d - 60) / 22
      img:drawPixel(x, y, blendC(DEEP_NAVY, DARK_BLUE, normD * 0.3))

    elseif d <= 60 then
      -- FACE CENTER with subtle gradient
      local normD = d / 60
      img:drawPixel(x, y, blendC(DEEP_NAVY, MED_BLUE, normD * 0.12))
    end
  end
end

-- ── DECORATIVE INNER RING ──
for y = 0, 191 do
  for x = 0, 191 do
    local d = dist(x, y)
    if d <= 62 and d > 59 then
      img:drawPixel(x, y, MED_BLUE)
    end
  end
end

-- ── DEGREE TICK MARKS (upper semicircle: -90 to 90) ──
for deg = -90, 90, 5 do
  local rad = (deg - 90) * PI / 180
  local isMajor = (deg % 45 == 0)
  local isMedium = (deg % 15 == 0)

  local r1 = 80
  local r2
  if isMajor then r2 = 62
  elseif isMedium then r2 = 68
  else r2 = 74 end

  local tickColor = WHITE
  if not isMajor and not isMedium then tickColor = LIGHT_GREY end

  for t = 0, 100 do
    local frac = t / 100
    local r = r1 + (r2 - r1) * frac
    local px = math.floor(cx + math.cos(rad) * r)
    local py = math.floor(cy + math.sin(rad) * r)
    if px >= 0 and px < 192 and py >= 0 and py < 192 then
      img:drawPixel(px, py, tickColor)
      if isMajor then
        -- Thicker
        local px2 = math.floor(cx + math.cos(rad + 0.008) * r)
        local py2 = math.floor(cy + math.sin(rad + 0.008) * r)
        if px2 >= 0 and px2 < 192 and py2 >= 0 and py2 < 192 then
          img:drawPixel(px2, py2, tickColor)
        end
        local px3 = math.floor(cx + math.cos(rad - 0.008) * r)
        local py3 = math.floor(cy + math.sin(rad - 0.008) * r)
        if px3 >= 0 and px3 < 192 and py3 >= 0 and py3 < 192 then
          img:drawPixel(px3, py3, tickColor)
        end
      end
    end
  end
end

-- ── Minor ticks on bottom half (subtle) ──
for deg = -85, 85, 15 do
  local rad = (deg + 90) * PI / 180
  for t = 0, 100 do
    local frac = t / 100
    local r = 80 + (76 - 80) * frac
    local px = math.floor(cx + math.cos(rad) * r)
    local py = math.floor(cy + math.sin(rad) * r)
    if px >= 0 and px < 192 and py >= 0 and py < 192 then
      img:drawPixel(px, py, DARK_GREY)
    end
  end
end

-- ── DEGREE LABELS at major positions ──
-- 3x5 pixel font for digits
local digitData = {
  ["0"] = {"111","101","101","101","111"},
  ["4"] = {"101","101","111","001","001"},
  ["5"] = {"111","100","111","001","111"},
  ["9"] = {"111","101","111","001","111"},
  ["-"] = {"000","000","111","000","000"},
}

local function drawChar(ox, oy, ch, color)
  local data = digitData[ch]
  if not data then return end
  for row = 1, #data do
    local line = data[row]
    for col = 1, #line do
      if line:sub(col, col) == "1" then
        local px = ox + col - 1
        local py = oy + row - 1
        if px >= 0 and px < 192 and py >= 0 and py < 192 then
          img:drawPixel(px, py, color)
        end
      end
    end
  end
end

local function drawString(ox, oy, str, color)
  for i = 1, #str do
    drawChar(ox + (i-1) * 4, oy, str:sub(i, i), color)
  end
end

-- Place labels at key degree positions
local labels = {
  {deg=-90, text="-90"},
  {deg=-45, text="-45"},
  {deg=0, text="0"},
  {deg=45, text="45"},
  {deg=90, text="90"},
}
for _, lbl in ipairs(labels) do
  local rad = (lbl.deg - 90) * PI / 180
  local r = 50
  local lx = math.floor(cx + math.cos(rad) * r)
  local ly = math.floor(cy + math.sin(rad) * r)
  local textW = #lbl.text * 4
  drawString(lx - math.floor(textW / 2), ly - 2, lbl.text, LIGHT_GREY)
end

-- ── "N" LABEL (RED) at top ──
local nData = {
  "10001",
  "11001",
  "10101",
  "10011",
  "10001",
}
local nx = cx - 2
local ny = cy - 55
for row = 1, #nData do
  local line = nData[row]
  for col = 1, #line do
    if line:sub(col, col) == "1" then
      img:drawPixel(nx + col - 1, ny + row - 1, RED)
    end
  end
end

-- ── "S" LABEL (BLUE) at bottom ──
local sData = {
  "01110",
  "10000",
  "01100",
  "00010",
  "11100",
}
local sx2 = cx - 2
local sy2 = cy + 51
for row = 1, #sData do
  local line = sData[row]
  for col = 1, #line do
    if line:sub(col, col) == "1" then
      img:drawPixel(sx2 + col - 1, sy2 + row - 1, BRIGHT_BLUE)
    end
  end
end

-- ── CENTER PIVOT ──
for y2 = -5, 5 do
  for x2 = -5, 5 do
    local d = math.sqrt(x2*x2 + y2*y2)
    if d <= 5 and d > 3 then
      img:drawPixel(cx + x2, cy + y2, BRASS)
    elseif d <= 3 and d > 1 then
      img:drawPixel(cx + x2, cy + y2, BRASS_LIGHT)
    elseif d <= 1 then
      img:drawPixel(cx + x2, cy + y2, WHITE)
    end
  end
end

-- ── GLASS REFLECTION ──
-- Upper-left reflection arc
for y = 0, 191 do
  for x = 0, 191 do
    local d = dist(x, y)
    if d <= 58 and d > 48 then
      local angle = myatan2(y - cy, x - cx)
      if angle > -2.6 and angle < -1.4 then
        -- Soft white highlight
        local existing = img:getPixel(x, y)
        local ec = Color(existing)
        if ec.alpha > 0 then
          local highlight = blendC(ec, WHITE, 0.12)
          img:drawPixel(x, y, highlight)
        end
      end
    end
  end
end

-- Small bright specular dot
for y2 = -2, 2 do
  for x2 = -2, 2 do
    local d = math.sqrt(x2*x2 + y2*y2)
    if d <= 2 then
      local px = cx - 25 + x2
      local py = cy - 30 + y2
      if px >= 0 and px < 192 and py >= 0 and py < 192 then
        img:drawPixel(px, py, blendC(WHITE, CYAN, 0.4))
      end
    end
  end
end

-- ── BEZEL SCREWS ──
local screws = {
  {cx, cy - 85}, {cx, cy + 85},
  {cx - 60, cy - 60}, {cx + 60, cy - 60},
  {cx - 60, cy + 60}, {cx + 60, cy + 60},
}
for _, pos in ipairs(screws) do
  local scx, scy = pos[1], pos[2]
  local d = dist(scx, scy)
  if d < 89 and d > 82 then
    for dy = -2, 2 do
      for dx = -2, 2 do
        local dd = math.sqrt(dx*dx + dy*dy)
        if dd <= 2 then
          local px = scx + dx
          local py = scy + dy
          if px >= 0 and px < 192 and py >= 0 and py < 192 then
            if dd <= 1 then
              img:drawPixel(px, py, BRASS_BRIGHT)
            else
              img:drawPixel(px, py, BRASS_DARK)
            end
          end
        end
      end
    end
    -- Screw slot line
    if scx >= 1 and scx < 191 then
      img:drawPixel(scx - 1, scy, BRASS_DARK)
      img:drawPixel(scx, scy, BRASS_DARK)
      img:drawPixel(scx + 1, scy, BRASS_DARK)
    end
  end
end

app.command.FlattenLayers()
spr:saveAs("C:\\Users\\shlom\\projects\\DNA Game\\assets\\sprites\\galvanometer.png")
app.command.CloseFile()
